In [97]:
import pandas as pd

In [98]:
matches = pd.read_csv('matches.csv', index_col=0)
matches.head()

date   time            comp        round  day venue result   gf   ga  \
1  2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2  2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3  2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   
4  2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away      W  1.0  0.0   
6  2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...  match report  notes    sh   sot  dist   fk   pk pkatt  \
1       Tottenham  ...  Match Report    NaN  18.0   4.0  16.9  1.0  0.0   0.0   
2    Norwich City  ...  Match Report    NaN  16.0   4.0  17.3  1.0  0.0   0.0   
3         Arsenal  ...  Match Report    NaN  25.0  10.0  14.3  0.0  0.0   0.0   
4  Leicester City  ...  Match Report    NaN  25.0   8.0  14.0  0.0  0.0   0.0   
6     Southampton  ...  Match Report    NaN  16.0   1.0  15.7  1.0  0.0   0.0   

   season             team  
1    2022  Manchester City  
2    2022  Manchester City  
3    2022  Manchester City  
4    2022  Manchester City  
6    2022  Manchester City  

[5 rows x 27 columns]

In [99]:
matches.shape

(1389, 27)

In [182]:
checkData = 38 * 20 * 2
print(checkData)
print(checkData - matches.shape[0])

1520
131


In the EPL, there are 38 matches in a season. There are 20 teams in a season. This data is only from 2 seasons.
If we look at the checkData - matches.shape[0] we see that there are 131 rows missing. So we're clearly missing some data.


In [102]:
matches["team"].value_counts()

team
Southampton                 72
Brighton and Hove Albion    72
Manchester United           72
West Ham United             72
Newcastle United            72
Burnley                     71
Leeds United                71
Crystal Palace              71
Manchester City             71
Wolverhampton Wanderers     71
Tottenham Hotspur           71
Arsenal                     71
Leicester City              70
Chelsea                     70
Aston Villa                 70
Everton                     70
Liverpool                   38
Fulham                      38
West Bromwich Albion        38
Sheffield United            38
Brentford                   34
Watford                     33
Norwich City                33
Name: count, dtype: int64

In the EPL, 3 teams get relegated each year and 3 teams get promoted so we should have 6 teams with fewer than 70 matches. But instead we have 7 teams with fewer than 70 matches and Liverpool was not relegated in the 202-2021 season. This is good information to have because now we know that Liverpool is missing one seasons worth of data.


In [103]:
matches["round"].value_counts()


round
Matchweek 1     39
Matchweek 16    39
Matchweek 34    39
Matchweek 32    39
Matchweek 31    39
Matchweek 29    39
Matchweek 28    39
Matchweek 26    39
Matchweek 25    39
Matchweek 24    39
Matchweek 23    39
Matchweek 2     39
Matchweek 19    39
Matchweek 17    39
Matchweek 20    39
Matchweek 15    39
Matchweek 5     39
Matchweek 3     39
Matchweek 13    39
Matchweek 12    39
Matchweek 4     39
Matchweek 11    39
Matchweek 10    39
Matchweek 9     39
Matchweek 8     39
Matchweek 14    39
Matchweek 7     39
Matchweek 6     39
Matchweek 30    37
Matchweek 27    37
Matchweek 22    37
Matchweek 21    37
Matchweek 18    37
Matchweek 33    32
Matchweek 35    20
Matchweek 36    20
Matchweek 37    20
Matchweek 38    20
Name: count, dtype: int64

In [104]:
# clean the data
matches.dtypes 


date             object
time             object
comp             object
round            object
day              object
venue            object
result           object
gf              float64
ga              float64
opponent         object
xg              float64
xga             float64
poss            float64
attendance      float64
captain          object
formation        object
referee          object
match report     object
notes           float64
sh              float64
sot             float64
dist            float64
fk              float64
pk              float64
pkatt           float64
season            int64
team             object
dtype: object

In [105]:
matches["date"] = pd.to_datetime(matches["date"])
matches.dtypes 

date            datetime64[ns]
time                    object
comp                    object
round                   object
day                     object
venue                   object
result                  object
gf                     float64
ga                     float64
opponent                object
xg                     float64
xga                    float64
poss                   float64
attendance             float64
captain                 object
formation               object
referee                 object
match report            object
notes                  float64
sh                     float64
sot                    float64
dist                   float64
fk                     float64
pk                     float64
pkatt                  float64
season                   int64
team                    object
dtype: object

In [106]:
##converting all objects to int or float to be processed by the machine learning software
matches["date"] = pd.to_datetime(matches["date"])
matches["h/a"] = matches["venue"].astype("category").cat.codes ## converting venue to a home (1) or away (0) number
matches["opp"] = matches["opponent"].astype("category").cat.codes ## converting opponents to a number
matches["hour"] = matches["time"].str.replace(":.+", "", regex=True).astype("int") ## converting hours to number in case a team plays better at a certain time
matches["day"] = matches["date"].dt.dayofweek ## converting day of week of game to a number

In [107]:
## setting a win to the value 1
matches["target"] = (matches["result"] == "W").astype("int") 
matches

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1    6  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2    5  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3    5  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4    5  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5    5  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34    6  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35    5  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36    6  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37    2  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38    6  Home      W  1.0   

     ga        opponent  ...  dist   fk   pk  pkatt season              team  \
1   1.0       Tottenham  ...  16.9  1.0  0.0    0.0   2022   Manchester City   
2   0.0    Norwich City  ...  17.3  1.0  0.0    0.0   2022   Manchester City   
3   0.0         Arsenal  ...  14.3  0.0  0.0    0.0   2022   Manchester City   
4   0.0  Leicester City  ...  14.0  0.0  0.0    0.0   2022   Manchester City   
6   0.0     Southampton  ...  15.7  1.0  0.0    0.0   2022   Manchester City   
..  ...             ...  ...   ...  ...  ...    ...    ...               ...   
38  4.0       Tottenham  ...  17.4  0.0  0.0    0.0   2021  Sheffield United   
39  2.0  Crystal Palace  ...  11.4  1.0  0.0    0.0   2021  Sheffield United   
40  0.0         Everton  ...  17.0  0.0  0.0    0.0   2021  Sheffield United   
41  1.0   Newcastle Utd  ...  16.0  1.0  0.0    0.0   2021  Sheffield United   
42  0.0         Burnley  ...  17.0  0.0  0.0    0.0   2021  Sheffield United   

   h/a opp  hour  target  
1    0  18    16       0  
2    1  15    15       1  
3    1   0    12       1  
4    0  10    15       1  
6    1  17    15       0  
..  ..  ..   ...     ...  
38   0  18    19       0  
39   1   6    15       0  
40   0   7    19       1  
41   0  14    18       0  
42   1   4    16       1  

[1389 rows x 31 columns]

In [108]:
from sklearn.ensemble import RandomForestClassifier

# training 100 decision trees on the data with 10 samples in each tree
rf = RandomForestClassifier(n_estimators=100, min_samples_split=10, random_state=1)

In [109]:
train = matches[matches["date"] < "2022-01-01"]
test = matches[matches["date"] > "2022-01-01"]

Here we're training the model so that when we test it on data in the future, it has a better chance of predicting the outcome of the game because of the model having been trained.


In [110]:
predictors = ["opp", "h/a", "day", "hour"]
rf.fit(train[predictors], train["target"])

RandomForestClassifier(min_samples_split=10, random_state=1)

In [111]:
preds = rf.predict(test[predictors])

In [112]:
from sklearn.metrics import accuracy_score
acc = accuracy_score(test["target"], preds) ## testing accuracy

In [113]:
acc

0.605072463768116

In [114]:
combined = pd.DataFrame(dict(actual=test["target"], prediction=preds))
pd.crosstab(index=combined["actual"], columns=combined["prediction"])

prediction    0   1
actual             
0           142  30
1            79  25

Here we see that when we predicted a loss/draw, we were correct 142 times and wrong 79 times which is good for this accuracy_score based off 100 decision trees.

But when we predicted a win, we were correct 30 times and wrong 25 times which is almost 50% which is too close for us to be confident in the model.

Thus we'll use precision_score increase the accuracy of the model.

In [115]:
from sklearn.metrics import precision_score
precision_score(test["target"], preds)

0.45454545454545453

We'll create more predictors to improve the accuracy of the model.

Let's split our matches dataframe by teams because we want to compute rolling averages for each team which will allow us to see how well a team has been doing in their past games based on shot attempts, goals, etc.


In [167]:
grouped_matches = matches.groupby("team")
group = grouped_matches.get_group("Chelsea")
group.head()


date   time            comp        round  day venue result   gf   ga  \
1 2021-08-14  15:00  Premier League  Matchweek 1    5  Home      W  3.0  0.0   
2 2021-08-22  16:30  Premier League  Matchweek 2    6  Away      W  2.0  0.0   
3 2021-08-28  17:30  Premier League  Matchweek 3    5  Away      D  1.0  1.0   
4 2021-09-11  17:30  Premier League  Matchweek 4    5  Home      W  3.0  0.0   
6 2021-09-19  16:30  Premier League  Matchweek 5    6  Away      W  3.0  0.0   

         opponent  ...  dist   fk   pk  pkatt season     team h/a opp  hour  \
1  Crystal Palace  ...  20.8  4.0  0.0    0.0   2022  Chelsea   1   6    15   
2         Arsenal  ...  14.6  0.0  0.0    0.0   2022  Chelsea   0   0    16   
3       Liverpool  ...  12.3  0.0  0.0    0.0   2022  Chelsea   0  11    17   
4     Aston Villa  ...  22.5  0.0  0.0    0.0   2022  Chelsea   1   1    17   
6       Tottenham  ...  13.2  0.0  0.0    0.0   2022  Chelsea   0  18    16   

   target  
1       1  
2       1  
3       0  
4       1  
6       1  

[5 rows x 31 columns]

We need to compute the rolling averages for Liverpool and basically say that if we're in matchweek 4, how did Liverpool do in the past 3 weeks.


In [168]:
def rolling_averages(group, cols, new_cols):
    group = group.sort_values("date")
    rolling_stats = group[cols].rolling(3, closed="left").mean()
    group[new_cols] = rolling_stats
    group = group.dropna(subset=new_cols)
    return group


In [169]:
cols = ["gf", "ga", "sh", "sot", "dist", "fk", "pk", "pkatt"]
new_cols = [f"{c}_rolling" for c in cols]
new_cols


['gf_rolling',
 'ga_rolling',
 'sh_rolling',
 'sot_rolling',
 'dist_rolling',
 'fk_rolling',
 'pk_rolling',
 'pkatt_rolling']

In [170]:
rolling_averages(group, cols, new_cols)


date   time            comp         round  day venue result   gf  \
5  2020-10-03  12:30  Premier League   Matchweek 4    5  Home      W  4.0   
6  2020-10-17  15:00  Premier League   Matchweek 5    5  Home      D  3.0   
8  2020-10-24  17:30  Premier League   Matchweek 6    5  Away      D  0.0   
10 2020-10-31  15:00  Premier League   Matchweek 7    5  Away      W  3.0   
12 2020-11-07  17:30  Premier League   Matchweek 8    5  Home      W  4.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
44 2022-03-13  14:00  Premier League  Matchweek 29    6  Home      W  1.0   
47 2022-04-02  15:00  Premier League  Matchweek 31    5  Home      L  1.0   
49 2022-04-09  15:00  Premier League  Matchweek 32    5  Away      W  6.0   
52 2022-04-20  19:45  Premier League  Matchweek 25    2  Home      L  2.0   
53 2022-04-24  14:00  Premier League  Matchweek 34    6  Home      W  1.0   

     ga        opponent  ...  hour  target  gf_rolling  ga_rolling sh_rolling  \
5   0.0  Crystal Palace  ...    12       1    2.000000    2.000000  12.333333   
6   3.0     Southampton  ...    15       0    2.333333    1.666667  14.333333   
8   0.0  Manchester Utd  ...    17       0    3.333333    2.000000  17.000000   
10  0.0         Burnley  ...    15       1    2.333333    1.000000  11.000000   
12  1.0   Sheffield Utd  ...    17       1    2.000000    1.000000  10.666667   
..  ...             ...  ...   ...     ...         ...         ...        ...   
44  0.0   Newcastle Utd  ...    14       1    2.666667    0.333333  12.000000   
47  4.0       Brentford  ...    15       0    2.666667    0.333333  11.333333   
49  0.0     Southampton  ...    15       1    1.666667    1.666667  14.666667   
52  4.0         Arsenal  ...    19       0    2.666667    1.333333  17.666667   
53  0.0        West Ham  ...    14       1    3.000000    2.666667  18.666667   

   sot_rolling dist_rolling fk_rolling  pk_rolling  pkatt_rolling  
5     5.666667    20.366667   0.666667    0.333333       0.666667  
6     5.666667    18.933333   0.666667    0.666667       1.000000  
8     6.666667    15.300000   0.666667    0.666667       0.666667  
10    3.333333    15.300000   0.000000    0.666667       0.666667  
12    5.000000    15.733333   0.333333    0.000000       0.000000  
..         ...          ...        ...         ...            ...  
44    5.000000    15.600000   0.666667    0.000000       0.000000  
47    5.000000    15.133333   0.666667    0.000000       0.000000  
49    6.000000    16.100000   0.666667    0.000000       0.000000  
52    8.333333    17.000000   0.333333    0.000000       0.000000  
53    8.000000    17.133333   0.000000    0.000000       0.000000  

[67 rows x 39 columns]

In [171]:
matches_rolling = matches.groupby("team").apply(lambda x: rolling_averages(x, cols, new_cols))
matches_rolling = matches_rolling.droplevel('team') ## dropping extra index level
matches_rolling.index = range(matches_rolling.shape[0]) ## adding new index
matches_rolling


C:\Users\tahas\AppData\Local\Temp\ipykernel_22068\415310982.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  matches_rolling = matches.groupby("team").apply(lambda x: rolling_averages(x, cols, new_cols))


date   time            comp         round  day venue result   gf  \
0    2020-10-04  14:00  Premier League   Matchweek 4    6  Home      W  2.0   
1    2020-10-17  17:30  Premier League   Matchweek 5    5  Away      L  0.0   
2    2020-10-25  19:15  Premier League   Matchweek 6    6  Home      L  0.0   
3    2020-11-01  16:30  Premier League   Matchweek 7    6  Away      W  1.0   
4    2020-11-08  19:15  Premier League   Matchweek 8    6  Home      L  0.0   
...         ...    ...             ...           ...  ...   ...    ...  ...   
1312 2022-03-13  14:00  Premier League  Matchweek 29    6  Away      W  1.0   
1313 2022-03-18  20:00  Premier League  Matchweek 30    4  Home      L  2.0   
1314 2022-04-02  15:00  Premier League  Matchweek 31    5  Home      W  2.0   
1315 2022-04-08  20:00  Premier League  Matchweek 32    4  Away      L  0.0   
1316 2022-04-24  14:00  Premier League  Matchweek 34    6  Away      L  0.0   

       ga         opponent  ...  hour  target  gf_rolling  ga_rolling  \
0     1.0    Sheffield Utd  ...    14       1    2.000000    1.333333   
1     1.0  Manchester City  ...    17       0    1.666667    1.666667   
2     1.0   Leicester City  ...    19       0    1.000000    1.666667   
3     0.0   Manchester Utd  ...    16       1    0.666667    1.000000   
4     3.0      Aston Villa  ...    19       0    0.333333    0.666667   
...   ...              ...  ...   ...     ...         ...         ...   
1312  0.0          Everton  ...    14       1    1.333333    1.000000   
1313  3.0     Leeds United  ...    20       0    1.666667    0.666667   
1314  1.0      Aston Villa  ...    15       1    2.333333    1.000000   
1315  1.0    Newcastle Utd  ...    20       0    1.666667    1.333333   
1316  1.0          Burnley  ...    14       0    1.333333    1.666667   

     sh_rolling sot_rolling dist_rolling fk_rolling  pk_rolling  pkatt_rolling  
0      7.666667    3.666667    14.733333   0.666667    0.000000       0.000000  
1      5.333333    3.666667    15.766667   0.000000    0.000000       0.000000  
2      7.000000    3.666667    16.733333   0.666667    0.000000       0.000000  
3      9.666667    4.000000    16.033333   1.000000    0.000000       0.000000  
4      9.666667    2.666667    18.033333   1.000000    0.333333       0.333333  
...         ...         ...          ...        ...         ...            ...  
1312  12.333333    3.666667    19.300000   0.000000    0.000000       0.000000  
1313  12.333333    4.333333    19.600000   0.000000    0.000000       0.000000  
1314  13.000000    5.333333    19.833333   0.000000    0.000000       0.000000  
1315  13.000000    5.000000    18.533333   0.000000    0.000000       0.000000  
1316  10.000000    4.666667    17.633333   0.000000    0.000000       0.000000  

[1317 rows x 39 columns]

Now that we have new predictors, we can train the model again.


In [172]:
def make_predictions(data, predictors, model):
    train = data[data["date"] < "2022-01-01"]
    test = data[data["date"] > "2022-01-01"]
    rf.fit(train[predictors], train["target"])
    preds = rf.predict(test[predictors])
    combined = pd.DataFrame(dict(actual = test["target"], prediction=preds))
    precision = precision_score(test["target"], preds)
    return combined, precision

In [173]:
combined, precision = make_predictions(matches_rolling, predictors + new_cols, rf)
precision

0.5740740740740741

In [174]:
combined

actual  prediction
55         0           1
56         1           1
57         1           0
58         1           1
59         1           1
...      ...         ...
1312       1           0
1313       0           0
1314       1           0
1315       0           1
1316       0           0

[276 rows x 2 columns]

Our precision scored has increased from 45% to 57% which is a good improvement.

But there are still come mispredictions as we can see from the combined but it doesn't tell us which team actually played in each match meaning we're unable to see if we're mispredicting and team.

To fix that we'll add that information to the combined dataframe with combined.merge


In [175]:
combined = combined.merge(matches_rolling[["date", "team", "opponent", "result"]], left_index = True, right_index = True)
combined

actual  prediction       date                     team       opponent  \
55         0           1 2022-01-23                  Arsenal        Burnley   
56         1           1 2022-02-10                  Arsenal         Wolves   
57         1           0 2022-02-19                  Arsenal      Brentford   
58         1           1 2022-02-24                  Arsenal         Wolves   
59         1           1 2022-03-06                  Arsenal        Watford   
...      ...         ...        ...                      ...            ...   
1312       1           0 2022-03-13  Wolverhampton Wanderers        Everton   
1313       0           0 2022-03-18  Wolverhampton Wanderers   Leeds United   
1314       1           0 2022-04-02  Wolverhampton Wanderers    Aston Villa   
1315       0           1 2022-04-08  Wolverhampton Wanderers  Newcastle Utd   
1316       0           0 2022-04-24  Wolverhampton Wanderers        Burnley   

     result  
55        D  
56        W  
57        W  
58        W  
59        W  
...     ...  
1312      W  
1313      L  
1314      W  
1315      L  
1316      L  

[276 rows x 6 columns]

We noticed that the team name and the opponent name are not listed as the same. For example, team = Wolverhampton Wanderers and opponent = Wolves.

To make this consistent, we can use a dictionary and then the pandas map function with that dictionary.


In [176]:
class MissingDict(dict): ## creating a class that inherits from the dictionary class
    __missing__ = lambda self, key: key ## in case a team name is missing

map_values = {
    "Brighton and Hove Albion": "Brighton",
    "Manchester United": "Manchester Utd",
    "Tottenham Hotspur": "Tottenham", 
    "West Ham United": "West Ham", 
    "Wolverhampton Wanderers": "Wolves"
}

mapping = MissingDict(**map_values)
mapping["West Ham United"]


'West Ham'

We can use these new and fixed names in the pandas map method.

In [177]:
combined["new_team"] = combined["team"].map(mapping)
combined

actual  prediction       date                     team       opponent  \
55         0           1 2022-01-23                  Arsenal        Burnley   
56         1           1 2022-02-10                  Arsenal         Wolves   
57         1           0 2022-02-19                  Arsenal      Brentford   
58         1           1 2022-02-24                  Arsenal         Wolves   
59         1           1 2022-03-06                  Arsenal        Watford   
...      ...         ...        ...                      ...            ...   
1312       1           0 2022-03-13  Wolverhampton Wanderers        Everton   
1313       0           0 2022-03-18  Wolverhampton Wanderers   Leeds United   
1314       1           0 2022-04-02  Wolverhampton Wanderers    Aston Villa   
1315       0           1 2022-04-08  Wolverhampton Wanderers  Newcastle Utd   
1316       0           0 2022-04-24  Wolverhampton Wanderers        Burnley   

     result new_team  
55        D  Arsenal  
56        W  Arsenal  
57        W  Arsenal  
58        W  Arsenal  
59        W  Arsenal  
...     ...      ...  
1312      W   Wolves  
1313      L   Wolves  
1314      W   Wolves  
1315      L   Wolves  
1316      L   Wolves  

[276 rows x 7 columns]

Look at the dataframe and see that there is a team and the opponent they play. then the opponent versus the team. 
So basically x versus y and y versus x are in the dataframe entailing why it's best to merge it with itself.


In [178]:
merged = combined.merge(combined, left_on = ["date", "new_team"], right_on = ["date", "team"])
merged

actual_x  prediction_x       date   team_x       opponent_x result_x  \
0           0             1 2022-01-23  Arsenal          Burnley        D   
1           1             1 2022-02-10  Arsenal           Wolves        W   
2           1             0 2022-02-19  Arsenal        Brentford        W   
3           1             1 2022-02-24  Arsenal           Wolves        W   
4           1             1 2022-03-06  Arsenal          Watford        W   
..        ...           ...        ...      ...              ...      ...   
195         1             0 2022-03-13  Watford      Southampton        W   
196         0             0 2022-04-02  Watford        Liverpool        L   
197         0             0 2022-04-09  Watford     Leeds United        L   
198         0             0 2022-04-16  Watford        Brentford        L   
199         0             0 2022-04-23  Watford  Manchester City        L   

    new_team_x  actual_y  prediction_y   team_y       opponent_y result_y  \
0      Arsenal         0             1  Arsenal          Burnley        D   
1      Arsenal         1             1  Arsenal           Wolves        W   
2      Arsenal         1             0  Arsenal        Brentford        W   
3      Arsenal         1             1  Arsenal           Wolves        W   
4      Arsenal         1             1  Arsenal          Watford        W   
..         ...       ...           ...      ...              ...      ...   
195    Watford         1             0  Watford      Southampton        W   
196    Watford         0             0  Watford        Liverpool        L   
197    Watford         0             0  Watford     Leeds United        L   
198    Watford         0             0  Watford        Brentford        L   
199    Watford         0             0  Watford  Manchester City        L   

    new_team_y  
0      Arsenal  
1      Arsenal  
2      Arsenal  
3      Arsenal  
4      Arsenal  
..         ...  
195    Watford  
196    Watford  
197    Watford  
198    Watford  
199    Watford  

[200 rows x 13 columns]